File: demos/topic_31357/03/lib/ai-service.js
/**
 * AI服务集成模块 - 负责调用大模型API进行PRD智能生成
 * 支持自定义API地址，兼容OpenAI接口格式
 */

const DEFAULT_API_URL = 'https://api.openai.com/v1/chat/completions';
const DEFAULT_MODEL = 'gpt-4o-mini';

class AIService {
  constructor() {
    this.apiKey = '';
    this.apiUrl = '';
    this.model = '';
  }

  /**
   * 初始化配置
   */
  async init() {
    const configs = await chrome.storage.local.get(['apiKey', 'apiUrl', 'model']);
    this.apiKey = configs.apiKey || '';
    this.apiUrl = configs.apiUrl || DEFAULT_API_URL;
    this.model = configs.model || DEFAULT_MODEL;
  }

  /**
   * 更新配置 - 只更新传入的字段，未传入的字段保持不变
   */
  async updateConfig({ apiKey, apiUrl, model }) {
    if (apiKey !== undefined) this.apiKey = apiKey;
    if (apiUrl !== undefined) this.apiUrl = apiUrl || DEFAULT_API_URL;
    if (model !== undefined) this.model = model || DEFAULT_MODEL;

    // 只写入实际更新的字段，避免覆盖未传入的有效值
    const updateData = {};
    if (apiKey !== undefined) updateData.apiKey = this.apiKey;
    if (apiUrl !== undefined) updateData.apiUrl = this.apiUrl;
    if (model !== undefined) updateData.model = this.model;

    if (Object.keys(updateData).length > 0) {
      await chrome.storage.local.set(updateData);
    }
  }

  /**
   * 检查API Key是否已配置
   */
  isConfigured() {
    return !!this.apiKey && this.apiKey.trim().length > 0;
  }

  /**
   * 测试API连接
   */
  async testConnection() {
    if (!this.isConfigured()) {
      return { success: false, message: 'API Key未配置' };
    }
    try {
      const response = await fetch(this.apiUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model: this.model,
          messages: [{ role: 'user', content: '你好，请回复"连接成功"' }],
          max_tokens: 20,
        }),
      });
      if (response.ok) {
        const data = await response.json();
        if (data.choices && data.choices.length > 0) {
          return { success: true, message: '连接成功，API Key有效' };
        }
        return { success: false, message: 'API返回数据格式异常' };
      } else {
        const errorData = await response.json().catch(() => ({}));
        const errorMsg = errorData.error?.message || `HTTP错误 ${response.status}`;
        return { success: false, message: `连接失败：${errorMsg}` };
      }
    } catch (error) {
      return { success: false, message: `网络错误：${error.message}` };
    }
  }

  /**
   * 调用AI生成PRD文档
   */
  async generatePRD(pageData) {
    if (!this.isConfigured()) {
      throw new Error('请先配置有效的API Key');
    }

    const systemPrompt = this._buildSystemPrompt();
    const userPrompt = this._buildUserPrompt(pageData);

    try {
      const response = await fetch(this.apiUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model: this.model,
          messages: [
            { role: 'system', content: systemPrompt },
            { role: 'user', content: userPrompt },
          ],
          temperature: 0.3,
          max_tokens: 4096,
        }),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        const errorMsg = errorData.error?.message || `HTTP错误 ${response.status}`;
        if (response.status === 401) {
          throw new Error('API Key无效或已过期，请重新配置');
        }
        throw new Error(`AI接口调用失败：${errorMsg}`);
      }

      const data = await response.json();
      if (!data.choices || data.choices.length === 0) {
        throw new Error('AI返回数据为空，请重试');
      }

      return data.choices[0].message.content;
    } catch (error) {
      if (error.message.includes('API Key') || error.message.includes('AI接口')) {
        throw error;
      }
      throw new Error(`网络请求失败：${error.message}，请检查网络连接和API地址`);
    }
  }

  /**
   * 构建系统提示词
   */
  _buildSystemPrompt() {
    return `你是一位资深产品经理，擅长编写标准化、结构化的产品需求文档(PRD)。
你的任务是根据提供的网页交互数据和页面信息，生成一份完整的PRD文档。

要求：
1. 严格按照指定的Markdown模板结构输出
2. 内容必须基于提供的页面数据，不得臆造功能
3. 语言为简体中文，逻辑通顺、专业规范
4. 所有模块必须完整输出，不可省略
5. 输出纯Markdown格式，不要用代码块包裹

输出模板结构：
# 【页面功能名】产品需求文档

## 1. 功能概述
（当前页面/模块的业务用途、使用场景、核心价值）

## 2. 页面信息
- 页面URL：
- 页面名称：
- 页面类型：（列表页/详情页/编辑页/弹窗功能等）

## 3. 核心功能列表
| 序号 | 功能名称 | 功能描述 |
|------|---------|---------|
| 1    |         |         |

## 4. 操作流程说明
（用户操作步骤、功能触发逻辑、前置条件、后置结果）

## 5. 页面元素说明
### 5.1 按钮说明
### 5.2 表单说明
### 5.3 弹窗说明
### 5.4 筛选项说明

## 6. 业务规则
### 6.1 表单校验规则
### 6.2 状态限制规则
### 6.3 权限逻辑
### 6.4 特殊场景规则

## 7. 异常场景
| 序号 | 异常场景 | 触发条件 | 提示信息/处理方式 |
|------|---------|---------|-----------------|

## 8. 功能优先级
- 默认优先级：普通
- 说明：（如有特殊优先级标注）`;
  }

  /**
   * 构建用户提示词
   */
  _buildUserPrompt(pageData) {
    const { pageInfo, interactions, elements, url, title } = pageData;

    let prompt = `请根据以下网页数据生成PRD文档：\n\n`;
    prompt += `## 页面基础信息\n`;
    prompt += `- 页面标题：${title || '未知'}\n`;
    prompt += `- 页面URL：${url || '未知'}\n\n`;

    if (pageInfo && pageInfo.length > 0) {
      prompt += `## 页面结构信息\n`;
      prompt += pageInfo.map(info => `- ${info}`).join('\n');
      prompt += '\n\n';
    }

    if (elements && elements.length > 0) {
      prompt += `## 页面元素信息\n`;
      prompt += elements.map(el => {
        let desc = `- 类型：${el.type}，标签：${el.label || el.text || '无'}`;
        if (el.placeholder) desc += `，占位符：${el.placeholder}`;
        if (el.required) desc += `，必填`;
        if (el.options && el.options.length > 0) desc += `，选项：${el.options.join('/')}`;
        if (el.validation) desc += `，校验：${el.validation}`;
        return desc;
      }).join('\n');
      prompt += '\n\n';
    }

    if (interactions && interactions.length > 0) {
      prompt += `## 用户交互行为记录（按操作顺序）\n`;
      prompt += interactions.map((inter, i) => {
        let desc = `${i + 1}. 操作类型：${inter.type}`;
        if (inter.target) desc += `，目标：${inter.target}`;
        if (inter.value !== undefined && inter.value !== '') desc += `，输入值：${inter.value}`;
        if (inter.result) desc += `，结果：${inter.result}`;
        return desc;
      }).join('\n');
      prompt += '\n\n';
    }

    prompt += `请严格按照PRD模板结构生成完整的Markdown文档，确保所有模块内容完整、专业。`;
    return prompt;
  }
}

// 导出单例
const aiService = new AIService();

File: public/demos/topic_151479/evolution-mirror/backend/app/services/ai_service.py
"""
AI 调度服务
支持多个 LLM 提供商，统一使用 OpenAI 兼容 API 格式
"""

import json
import logging
import httpx
from typing import Optional

logger = logging.getLogger(__name__)

# 预设的提供商配置
PRESET_PROVIDERS = {
    "deepseek": {
        "name": "DeepSeek",
        "base_url": "https://api.deepseek.com/v1",
        "models": ["deepseek-chat", "deepseek-reasoner"],
        "default_model": "deepseek-chat",
    },
    "qwen": {
        "name": "通义千问",
        "base_url": "https://dashscope.aliyuncs.com/compatible-mode/v1",
        "models": ["qwen-turbo", "qwen-plus", "qwen-max"],
        "default_model": "qwen-turbo",
    },
    "ernie": {
        "name": "文心一言",
        "base_url": "https://qianfan.baidubce.com/v2",
        "models": ["ernie-speed-128k", "ernie-lite-8k", "ernie-4.0-8k"],
        "default_model": "ernie-speed-128k",
    },
}


class AIService:
    """AI 调度服务，管理 LLM API 调用"""

    def __init__(self):
        self._provider: str = "deepseek"
        self._base_url: str = PRESET_PROVIDERS["deepseek"]["base_url"]
        self._model: str = PRESET_PROVIDERS["deepseek"]["default_model"]
        self._api_key: str = ""
        self._enabled: bool = False

    @property
    def is_configured(self) -> bool:
        return self._enabled and bool(self._api_key)

    def configure(self, provider: str, api_key: str, model: Optional[str] = None, base_url: Optional[str] = None):
        """配置 AI 服务"""
        self._api_key = api_key
        self._enabled = True

        if provider in PRESET_PROVIDERS:
            self._provider = provider
            preset = PRESET_PROVIDERS[provider]
            self._base_url = base_url or preset["base_url"]
            self._model = model or preset["default_model"]
        else:
            # 自定义提供商
            self._provider = provider
            self._base_url = base_url or "https://api.openai.com/v1"
            self._model = model or "gpt-3.5-turbo"

    def get_config(self) -> dict:
        """获取当前配置"""
        return {
            "provider": self._provider,
            "base_url": self._base_url,
            "model": self._model,
            "api_key_set": bool(self._api_key),
            "enabled": self._enabled,
        }

    def get_presets(self) -> dict:
        """获取所有预设提供商信息（不包含 api_key）"""
        return PRESET_PROVIDERS

    async def chat(self, messages: list[dict], temperature: float = 0.7, max_tokens: int = 2000) -> str:
        """
        发送聊天请求

        Args:
            messages: 消息列表 [{"role": "system/user/assistant", "content": "..."}]
            temperature: 温度参数
            max_tokens: 最大 token 数

        Returns:
            AI 回复文本

        Raises:
            RuntimeError: AI 未配置或请求失败
        """
        if not self.is_configured:
            raise RuntimeError("AI 服务未配置，请先在设置中填入 API Key")

        url = f"{self._base_url}/chat/completions"
        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "Content-Type": "application/json",
        }
        payload = {
            "model": self._model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
        }

        try:
            async with httpx.AsyncClient(timeout=120.0) as client:
                resp = await client.post(url, headers=headers, json=payload)
                resp.raise_for_status()
                data = resp.json()
                return data["choices"][0]["message"]["content"]
        except httpx.TimeoutException:
            raise RuntimeError("AI 请求超时，请检查网络连接")
        except httpx.HTTPStatusError as e:
            detail = ""
            try:
                detail = e.response.json().get("error", {}).get("message", str(e))
            except Exception:
                detail = str(e)
            raise RuntimeError(f"AI 请求失败: {detail}")
        except Exception as e:
            raise RuntimeError(f"AI 请求异常: {e}")

    async def analyze_note(self, title: str, content: str) -> dict:
        """
        分析笔记内容，提取待办事项、时间信息、关键主题

        Returns:
            {"todos": [...], "schedule_items": [...], "topics": [...], "summary": "..."}
        """
        system_prompt = """你是一个个人助手，帮助用户分析笔记内容。请从用户的笔记中提取以下信息，以 JSON 格式返回：

{
  "todos": ["待办事项1", "待办事项2"],
  "schedule_items": [{"content": "日程内容", "datetime": "2024-01-15 14:00"}],
  "topics": ["主题1", "主题2"],
  "summary": "一句话摘要"
}

规则：
1. todos: 从文本中识别明确的待办、承诺、约定
2. schedule_items: 识别包含时间信息的条目，datetime 尽量提取原文中的时间表达
3. topics: 提取 1-5 个关键主题/话题
4. summary: 用一句话概括这条笔记的核心内容
5. 如果某项提取不到，返回空数组
6. 只返回 JSON，不要其他文字"""

        user_prompt = f"标题：{title}\n\n内容：{content}"

        try:
            result = await self.chat(
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt},
                ],
                temperature=0.3,
            )
            # 尝试解析 JSON
            result = result.strip()
            if result.startswith("```"):
                # 去掉 markdown 代码块
                lines = result.split("\n")
                result = "\n".join(lines[1:-1])
            return json.loads(result)
        except json.JSONDecodeError:
            return {"todos": [], "schedule_items": [], "topics": [], "summary": result}
        except RuntimeError:
            return {"todos": [], "schedule_items": [], "topics": [], "summary": "AI 分析不可用"}

    async def generate_tags(self, title: str, content: str) -> list[str]:
        """为笔记自动生成标签建议"""
        system_prompt = """为以下笔记生成 2-5 个标签，以 JSON 数组格式返回，如 ["标签1", "标签2"]。只返回 JSON 数组，不要其他文字。"""
        user_prompt = f"标题：{title}\n内容：{content}"

        try:
            result = await self.chat(
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt},
                ],
                temperature=0.5,
                max_tokens=200,
            )
            result = result.strip()
            if result.startswith("```"):
                lines = result.split("\n")
                result = "\n".join(lines[1:-1])
            tags = json.loads(result)
            return tags if isinstance(tags, list) else []
        except Exception:
            return []


# 全局单例
ai_service = AIService()